### PR TITLE
curation: add keyword api

### PR DESCRIPTION
**Pull request by MJedr**
Ref: #1676

### CODE REVIEW

File: backend/inspirehep/curation/views.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    datatype_warnings = check_keyword_dict_datatypes(json, required_fields)
+    if datatype_warnings:
+        return (
+            jsonify(
+                success=False,
+                message=f"""{" ".join(datatype_warnings)}""",
+            ),
+            400,
+        )
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:45]: this is not something you should be doing manually but use `webargs` for this (validate the payload based on marshmallow schema).

File: backend/inspirehep/curation/views.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    if desy_bookkeeping:
+        record_desy_bookkeeping = record.get("_desy_bookkeeping", [])
+        record_desy_bookkeeping.append(desy_bookkeeping)
+
+    builder = LiteratureBuilder()
+    for keywords in keywords:
+        builder.add_keyword(keywords)
+    if builder.record.get("keywords"):
+        record["keywords"] = builder.record.get("keywords")
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:59]: this is incorrect, you should remove all keywords with `schema: INSPIRE` (don't touch the others!) and add all of the provided keywords with no `source` and `schema: INSPIRE`.

File: backend/inspirehep/curation/views.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+@blueprint.route("/literature/<int:pid_value>/keywords", methods=["PUT"])
+@login_required_with_roles([Roles.cataloger.value])
+@parser.use_kwargs(
+    {
+        "keywords": fields.List(fields.String, required=False),
+        "_desy_bookkeeping": fields.Dict(required=False),
+    }
+)
+def add_keywords(pid_value, **kwargs):
```
Comment (drjova) [inline backend/inspirehep/curation/views.py:36]:
```suggestion
def add_keywords(pid_value, keywords, _desybookkeeping):
```

File: backend/inspirehep/curation/views.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+def handle_error(error, req, schema, error_status_code, error_headers):
+    message = f"""Incorrect input type for fields: {" ".join(error.field_names)}"""
+    abort(400, message)
+
+
+@blueprint.route("/literature/<int:pid_value>/keywords", methods=["PUT"])
+@login_required_with_roles([Roles.cataloger.value])
+@parser.use_kwargs(
+    {
+        "keywords": fields.List(fields.String, required=False),
```
Comment (drjova) [inline backend/inspirehep/curation/views.py:32]:
```suggestion
        "keywords": fields.List(fields.String, required=False, missing=[]),
```

File: backend/inspirehep/curation/views.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        "_desy_bookkeeping": fields.Dict(required=False),
```
Comment (drjova) [inline backend/inspirehep/curation/views.py:33]:
```suggestion
        "_desy_bookkeeping": fields.Dict(required=False, missing={}),
```

File: backend/inspirehep/curation/views.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+)
+def add_keywords(pid_value, **kwargs):
+    keywords = kwargs.get("keywords", [])
+    desy_bookkeeping = kwargs.get("_desy_bookkeeping", {})
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:38]: providing a default argument to `get` is useful in case you're going to process it further no matter what, but here you're doing explicit `if` checks later, so it's not really needed.

File: backend/inspirehep/curation/views.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+from webargs import fields
+from webargs.flaskparser import FlaskParser
+
+from inspirehep.accounts.decorators import login_required_with_roles
+from inspirehep.accounts.roles import Roles
+from inspirehep.records.api import LiteratureRecord
+from inspirehep.serializers import jsonify
+
+blueprint = Blueprint("inspirehep_curation", __name__, url_prefix="/curation")
+parser = FlaskParser()
```
Comment (drjova) [inline backend/inspirehep/curation/views.py:19]: why do we need this? `from webargs.flaskparser import parser` should be enough no?

File: backend/inspirehep/curation/views.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                message="None of required fields was passed",
+            ),
+            400,
+        )
+
+    record = LiteratureRecord.get_record_by_pid_value(pid_value)
+
+    if desy_bookkeeping:
+        record_desy_bookkeeping = record.get("_desy_bookkeeping", [])
+        record_desy_bookkeeping.append(desy_bookkeeping)
```
Comment (drjova) [inline backend/inspirehep/curation/views.py:53]: the record is never updated with these values, or I'm missing something?

File: backend/inspirehep/curation/views.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+from inspirehep.records.api import LiteratureRecord
+from inspirehep.serializers import jsonify
+
+blueprint = Blueprint("inspirehep_curation", __name__, url_prefix="/curation")
+parser = FlaskParser()
+
+
+@parser.error_handler
+def handle_error(error, req, schema, error_status_code, error_headers):
+    message = f"""Incorrect input type for fields: {" ".join(error.field_names)}"""
```
Comment (drjova) [inline backend/inspirehep/curation/views.py:24]:
```suggestion
    message = f"Incorrect input type for fields: {''.join(error.field_names)}"
```

File: backend/inspirehep/curation/views.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    abort(400, message)
+
+
+@blueprint.route("/literature/<int:pid_value>/keywords", methods=["PUT"])
+@login_required_with_roles([Roles.cataloger.value])
+@parser.use_kwargs(
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:30]: Minor: I think using `use_args` is a little bit more clear in this case, as you don't have `required` fields, so you need to use `**kwargs` (instead of `args`).

File: backend/inspirehep/curation/views.py (hunk 8 above)
Comment (michamos) [inline backend/inspirehep/curation/views.py:53]:
```suggestion
       record.setdefault("_desy_bookkeeping", []).append(desy_bookkeeping)
```
your version has a bug in case `_desy_bookkeeping` is not part of the record yet (you should maybe add a test for that): the first line will create an empty list, the second line will append to it, but that list is never assigned to the record.

File: backend/inspirehep/curation/views.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        "keywords": fields.List(fields.String, required=False),
+        "_desy_bookkeeping": fields.Dict(required=False),
+    }
+)
+def add_keywords(pid_value, **kwargs):
+    keywords = kwargs.get("keywords", [])
+    desy_bookkeeping = kwargs.get("_desy_bookkeeping", {})
+
+    if not any([keywords, desy_bookkeeping]):
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:40]: I would go for the explicit version here, as there are only 2 variables:
```suggestion
    if not keywords and not desy_bookkeeping:
```

File: backend/inspirehep/curation/views.py (hunk 7 above)
Comment (MJedr) [reply backend/inspirehep/curation/views.py:19]: I am overriding parser's `handle_error` method, it won't work without that

File: backend/inspirehep/curation/views.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+    if desy_bookkeeping:
+        record_desy_bookkeeping = record.get("_desy_bookkeeping", [])
+        record_desy_bookkeeping.append(desy_bookkeeping)
+
+    if keywords:
+        record_keywords = record.get("keywords", [])
+        for record_keyword in record_keywords:
+            if record_keyword.get("schema") == "INSPIRE":
+                record_keywords.remove(record_keyword)
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:59]: this is accidentally quadratic. Better do
```suggestion
    other_keywords = [keyword for keyword in record_keywords if keyword.get("schema") != "INSPIRE"]
    record["keywords"] = other_keywords
```
(and later append to `other_keywords` on L61).

File: backend/inspirehep/curation/views.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        for keyword in keywords:
+            record_keywords.append({"value": keyword, "schema": "INSPIRE"})
+
+    record.update(dict(record))
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:63]: what happens in case of validation error here (e.g. passing an empty string as one of the keywords, repeated keywords, wrong structure in `_desy_bookkeeping`, etc.)? I think that should be handled in order to display a nice error message rather than the generic error message for 500 internal error.

File: backend/pyproject.toml [hunk 14]
```diff
@@ -157,6 +157,7 @@ inspirehep_tools = "inspirehep.tools.views:blueprint"
 inspirehep_assign = "inspirehep.assign.views:blueprint"
 inspirehep_matcher = "inspirehep.matcher.views:blueprint"
 invenio_oaiserver = "invenio_oaiserver.views.server:blueprint"
+invenio_curation = "inspirehep.curation.views:blueprint"
```
Comment (michamos) [inline backend/pyproject.toml:160]:
```suggestion
inspirehep_curation = "inspirehep.curation.views:blueprint"
```

File: backend/inspirehep/curation/views.py (hunk 7 above)
Comment (michamos) [reply backend/inspirehep/curation/views.py:19]: why do you need to override `handle_error`?

Comment (MJedr) [reply backend/inspirehep/curation/views.py:19]: To return custom message and status code, without that it would be `422 Unprocessable Entity` without detailed message

File: backend/inspirehep/curation/views.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        ]
+        for keyword in keywords:
+            other_keywords.append({"value": keyword, "schema": "INSPIRE"})
+        record["keywords"] = other_keywords
+    try:
+        record.update(dict(record))
+        db.session.commit()
+        return jsonify(success=True)
+    except ValidationError as exception:
+        return jsonify(success=False, message=exception.message), 500
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:67]: I would say this is a 400 for a wrong request, not a generic server internal error.

**Review by michamos — APPROVED** [review]
minor comment, looks good otherwise :+1:

File: backend/inspirehep/curation/views.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+@parser.error_handler
+def handle_error(error, req, schema, error_status_code, error_headers):
+    message = f"Incorrect input type for fields: {''.join(error.field_names)}"
+    abort(400, message)
+
+
+@blueprint.route("/literature/<int:pid_value>/keywords", methods=["PUT"])
+@login_required_with_roles([Roles.cataloger.value])
+@parser.use_args(
+    {"_desy_bookkeeping": fields.Dict(), "keywords": fields.List(fields.String)},
```
Comment (michamos) [inline backend/inspirehep/curation/views.py:32]: don't you need the `required=False` any longer?
